evm/p2p/kademlia.py

```diff
@@ -0,0 +1,446 @@
+    most-recently seen at the tail. For small values of i, the k-buckets will generally
+    be empty (as no appropriate nodes will exist). For large values of i, the lists can
+    grow up to size k, where k is a system-wide replication parameter.
+    k is chosen such that any given k nodes are very unlikely to fail within an hour of
+    each other (for example k = 20).
+    """
+    k = k_bucket_size
+
+    def __init__(self, start, end):
+        self.start = start
+        self.end = end
+        self.nodes = []
+        self.replacement_cache = []
+        self.last_updated = time.time()
+
+    @property
+    def midpoint(self):
+        return self.start + (self.end - self.start) // 2
+
+    def id_distance(self, id):
+        return self.midpoint ^ id
+
+    def nodes_by_id_distance(self, id):
+        return sorted(self.nodes, key=operator.methodcaller('id_distance', id))
+
+    def split(self):
+        """Split at the median id"""
+        splitid = self.midpoint
+        lower = KBucket(self.start, splitid)
+        upper = KBucket(splitid + 1, self.end)
+        for node in self.nodes:
+            bucket = lower if node.id <= splitid else upper
+            bucket.add_node(node)
+        for node in self.replacement_cache:
+            bucket = lower if node.id <= splitid else upper
+            bucket.replacement_cache.append(node)
+        return lower, upper
+
+    def remove_node(self, node):
+        if node not in self.nodes:
+            return
+        self.nodes.remove(node)
+
+    def in_range(self, node):
+        return self.start <= node.id <= self.end
+
+    @property
+    def is_full(self):
+        return len(self) == self.k
+
+    def add_node(self, node):
+        """
+        If the sending node already exists in the recipient’s k-bucket,
+        the recipient moves it to the tail of the list.
+
+        If the node is not already in the appropriate k-bucket and the bucket has fewer than k
+        entries, then the recipient just inserts the new sender at the tail of the list.
+
+        If the  appropriate k-bucket is full, however, then the recipient pings the k-bucket’s
+        least-recently seen node to decide what to do.
+
+        on success: return None
+        on bucket full: return least recently seen Node for eviction check
+        """
+        self.last_updated = time.time()
+        if node in self.nodes:  # already exists
+            self.nodes.remove(node)
+            self.nodes.append(node)
+        elif len(self) < self.k:  # add if fewer than k entries
+            self.nodes.append(node)
+        else:  # bucket is full
+            return self.head
+
+    @property
+    def head(self):
+        """Least recently seen"""
+        return self.nodes[0]
+
+    @property
+    def depth(self):
+        """Depth is the prefix shared by all nodes in the bucket.
+
+        i.e. The number of shared leading bits.
+        """
+        def to_binary(x):  # left padded bit representation
```

Comment:
Can we pull this out to a utility somewhere (and test it!)

Reply:
I've just started working on unit tests for this stuff, and that is going to lead to many refactorings, very possibly what you suggest or maybe a complete rewrite